feat: Add rez versioning

skral opened the pull request:
## Changes:

Add [rez](https://github.com/nerdvegas/rez) versioning, including tests, and documentation.

## Context:

As described in discussion [10918](https://github.com/renovatebot/renovate/discussions/10918), I am using the rez package manager and can read the package.py it uses to state dependencies with a regular expression, the versioning needed some code related changes though. This MR adds the capability to renovate to understand rez versioning.

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number -->

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,13 @@
+rez versioning was developed to support the [rez Package Manager](https://github.com/nerdvegas/rez). It's based on Semantic versioning but includes its own concept of ranges.
```
```suggestion
rez versioning was developed to support the [rez Package Manager](https://github.com/nerdvegas/rez).
It's based on Semantic versioning but includes its own concept of ranges.
```

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+**Use of dots**
+
+A range can be expressed with dots `1.2..2`, it means `>=1.2.x <2.0.0`.
```
```suggestion
A range is expressed with dots `1.2..2` means `>=1.2.x <2.0.0`.
```

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+**No exact versions unless using two equals ==**
+
+In rez, `1.2.3` doesn't mean "exactly 1.2.3", it actually means `>= 1.2.3 <1.2.4`.
```
```suggestion
In rez, `1.2.3` doesn't mean "exactly 1.2.3", it actually means `>= 1.2.3 <1.2.4`.
If you want to use an exact version use two equal characters, like this: `==1.2.3`.
```

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+**Use of pipes**
+
+This has not been implemented yet. The current iteration of rez versioning does not support pipes yet.
```
```suggestion
This has not been implemented yet.
The current iteration of rez versioning does not support pipes yet.
```

HonkingGoose reviewed (changes requested):
Some initial docs changes.

viceice commented on lib/versioning/rez/index.spec.ts:
```diff
@@ -0,0 +1,583 @@
+    it('supports isVersion', () => {
+      expect(versioning.isVersion('1.2.3')).toBeTruthy();
+    });
+    it('understands rez version ranges', () => {
```
Please add newline before `it`, exerpt the first in `describe` block

viceice reviewed (commented):
Code otherwise LGTM so far.

viceice reviewed (changes requested):
Needs lint and coverage fixes

skral commented:
Applied the docs suggestions, updated the tests to reach 100% coverage and ran prettier. Please take another look when you get a chance.

HonkingGoose commented on lib/versioning/rez/index.ts:
```diff
@@ -0,0 +1,324 @@
+export const supportsRanges = true;
+export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+// Regular Expressions have been copied from, some more work were necessary to make it work:
+// https://github.com/nerdvegas/rez/blob/master/src/rez/vendor/version/version.py
```
License for the copied work is here:

https://github.com/nerdvegas/rez/blob/master/COPYING.LESSER

I don't know if copy/pasting and adjusting like this is allowed under our and their license?

viceice replied:
🤔 I'm not sure if a regex expression is protectable, but we should ask them to be sure.

nerdvegas/rez#1110

skral replied:
Very good question, did not think of that. @viceice, thanks for creating the issue officially right away.

Hope it's not a problem, would be unnecessarily hard to come up with different logic for this.

viceice replied:
We got an answer. Please add a clear comment that some regexes are from rez and have a GPL license.

@rarkins Is it ok to integrate that code?

rarkins replied:
Unfortunately it's not OK, unless we had a way to remove rez versioning from commercial editions downstream. WhiteSource Renovate On-Premises for example is EULA-licensed and although it's free, we can't bundle others GPL code within it and so could no longer offer it.

What we would need to do is have someone new to this - so unfortunately not @skral or any of us in this thread now - implement our own parsing based off a spec and completely independently of rez's own source code and without having seen the regexes. If our rez versioning is in no way derived from their GPL'd code, then there's less chance of a problem.

The alternative would be that whomever contributed to the regexes part of their code grants us a permissive license for their use but that's perhaps unlikely.

Final alternative: we finally implement a "plugin" architecture so that Renovate users can more easily add in their own custom versioning/managers/datasources, in which case rez could exist as a GPL'd Renovate plugin, easily combinable with Renovate in a custom build.

skral replied:
> The alternative would be that whomever contributed to the regexes part of their code grants us a permissive license for their use but that's perhaps unlikely.

The original author (i.e. nerdvegas - via git blame of the code section) of the regular expressions we're talking about has said "If all you need to do is grab some regexes, I would just do that, and include a comment linking to the rez project.". That's not a license but isn't this the ok that is needed to copy the expressions?

> implement our own parsing based off a spec and completely independently of rez's own source code and without having seen the regexes

That is of course doable, I can find someone to help with that but I might as well approach the problem completely different and scrap the expressions.

Let me know what you think is the best way forward.

rarkins replied:
I think the response you got was in the context of "Renovate is AGPL, can we use your GPL regexes?" and not in the context I described afterwards, which is that we have a requirement to have rights to use code in a non-GPL way. If we were to reuse the regexes in the way we need, the best way I can think of to do that is if the original author submits a (non-functional) PR here simply containing the required regex snippets and completing our CLA. This would provide us the grant we need to reuse the regex while not having any impact on the remainder of rez.

And to clarify:
- I don't think the original author is under any obligation at all to do this, it's purely their choice
- The reason they'd want to do this is because the benefits of improving the rez ecosystem outweigh any negatives of granting Renovate/WhiteSource rights to use the snippets how they like

skral replied:
Asked the original author for help with creating the PR [here](https://github.com/nerdvegas/rez/issues/1110#issuecomment-891271552). We'll see what he says.

skral replied:
We should be good here now. Thanks to @nerdvegas.

viceice commented on lib/versioning/rez/index.ts:
```diff
@@ -0,0 +1,324 @@
+    return input.replace('==', '=');
+  }
+  if (inclusiveBound.test(input)) {
+    return '>=' + input.replace('..', ' <');
```
Is it expected to only replace the first occurence of `..`

skral replied:
rez doesn't support multiple dot ranges. You can argue replacing only the first occurrence is good enough. Let me know how you want to proceed.

viceice replied:
leave as is

viceice commented on lib/versioning/rez/index.ts:
```diff
@@ -0,0 +1,324 @@
+    return input;
+  }
+  if (inclusiveBound.test(input)) {
+    return '>=' + input.replace('..', ', <');
```
Here too?

skral replied:
Same as https://github.com/renovatebot/renovate/pull/10930#discussion_r678948400

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,16 @@
+rez versioning was developed to support the [rez Package Manager](https://github.com/nerdvegas/rez).
+It's based on Semantic versioning but includes its own concept of ranges.
```
```suggestion
It's based on Semantic Versioning but includes its own concept of ranges.
```

If `rez` is based on the Semantic Versioning spec (with the exception of different range behavior). I think we should capitalize the whole phrase "Semantic Versioning", as that's how it's styled in the Semantic Versioning specification as found here: https://semver.org/

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,16 @@
+In rez, `1.2.3` doesn't mean "exactly 1.2.3", it actually means `>= 1.2.3 <1.2.4`.
+If you want to use an exact version use two equal characters, like this: `==1.2.3`.
+
+**Use of pipes**
```
Do we need an explanation of what `pipes` are used for in rez, or is this obvious to people who use `rez`?
From reading the current doc we have here, it's not clear what functionality I'm missing because we lack pipe support.

skral replied:
How about adding something like this? Always better to make things easier for a new reader.

```suggestion
**Use of pipes**

rez uses pipes as an OR operator, "2.7..3|4" means ">=2.7 <3 OR 4.x.x".
```

HonkingGoose replied:
I like it! 👍

viceice commented:
Marking this as draft until we know licensing for sure

HonkingGoose commented on lib/versioning/rez/readme.md:
```diff
@@ -0,0 +1,18 @@
+
+**Use of pipes**
+
+rez uses pipes as an OR operator, "2.7..3|4" means ">=2.7 <3 OR 4.x.x".
```
```suggestion
rez uses pipes as an OR operator, `2.7..3|4` means `>=2.7 <3 OR 4.x.x`.
```

I think this will look better when using backticks.

skral replied:
Damn, obviously yes. Should have stuck to the existing style anyway. Thanks for noticing. Applied.

skral commented:
I have a question regarding documenting how to make renovate work with rez. This MR only updates the versioning but it won't be clear to a new user how to get renovate to read the file that defines dependencies for rez. I have been setting this up with regular expressions because to get renovate to work with rez I hoped I could do it all with just regular expressions and not do any code changes, versioning was the moment I had to do code changes.

My question: Is there a place I could document how to set up a configuration to make renovate read versions from rez dependencies and do its job? I am still contemplating if I am eager to spend more time and add a rez manager as well but in the mean time it would be very helpful for others to see how to use renovate with rez and make onboarding as easy as possible.

I am asking it here because I am happy to add more documentation in this PR already. I am equally fine to open a separate issue and PR for the documentation improvements. Asking because I might as well add it here while we wait for the license situation to clear up.

@viceice @HonkingGoose What do you think?

nerdvegas commented:
https://github.com/renovatebot/renovate/pull/11631

(apologies for the delay, long story!)

HonkingGoose commented:
We need to wait until PR:

- https://github.com/renovatebot/renovate/pull/11634

gets merged until we can proceed.

HonkingGoose commented:
> My question: Is there a place I could document how to set up a configuration to make renovate read versions from rez dependencies and do its job? I am still contemplating if I am eager to spend more time and add a rez manager as well but in the mean time it would be very helpful for others to see how to use renovate with rez and make onboarding as easy as possible.
> 
> I am asking it here because I am happy to add more documentation in this PR already. I am equally fine to open a separate issue and PR for the documentation improvements. Asking because I might as well add it here while we wait for the license situation to clear up.

If you know how to get this setup with _just_ the versioning, you could add documentation to the `lib/versioning/rez/readme.md` file that exists already on this PR.

If/once you get the Rez manager working, you can add documentation on how the Rez manager works, and maybe edit/delete the "versioning workaround" for Rez?

viceice commented:
needs to revert the rez excludes from #11634

viceice commented:
merged, so this can continue now 🎉

skral commented on lib/versioning/rez/index.ts:
```diff
@@ -57,3 +67,317 @@
+  minSatisfyingVersion,
+  sortVersions,
+};
+export default api;
```
Fix newline at end of file

```suggestion
export default api;

```

skral commented:
> > My question: Is there a place I could document how to set up a configuration to make renovate read versions from rez dependencies and do its job? I am still contemplating if I am eager to spend more time and add a rez manager as well but in the mean time it would be very helpful for others to see how to use renovate with rez and make onboarding as easy as possible.
> > I am asking it here because I am happy to add more documentation in this PR already. I am equally fine to open a separate issue and PR for the documentation improvements. Asking because I might as well add it here while we wait for the license situation to clear up.
> 
> If you know how to get this setup with _just_ the versioning, you could add documentation to the `lib/versioning/rez/readme.md` file that exists already on this PR.
> 
> If/once you get the Rez manager working, you can add documentation on how the Rez manager works, and maybe edit/delete the "versioning workaround" for Rez?

Sounds good. I will prepare a separate PR for it so this one can move on. I am really eager to play around with this with docker in a production environment.

skral commented:
I reverted previous changes necessary to allow for the rez versioning to exist without it being completely ready. Please take another look. Linters and tests are passing locally.

viceice commented on lib/versioning/rez/index.ts:
```diff
@@ -57,3 +67,317 @@
+// - Replace (?P<...>) -> (?<...>)
+// - Replace ?(...) -> \k<...>
+// - Replace single \ -> double \
+const versionGroup = '([0-9a-zA-Z_]+(?:[.-][0-9a-zA-Z_]+)*)';
```
Please split the file as it's pretty big now. Maybe one file for regex and one for non public util function's.

skral replied:
Updated. Please have another look.

viceice commented on lib/versioning/rez/index.spec.ts:
```diff
@@ -0,0 +1,666 @@
+
+describe('versioning/rez/index', () => {
+  describe('equals', () => {
+    it.each([
```
@zharinov can you make suggestions for table style format?

viceice reviewed (commented):
Looks good so far, some minor tweak.

zharinov commented on lib/versioning/rez/index.spec.ts:
```diff
@@ -0,0 +1,666 @@
+      ['1.9.0', '1.9'],
+    ])('%s == %s', (a, b) => {
+      expect(versioning.equals(a, b)).toBe(true);
+    });
```
```ts
test.each`
  a             | b        | expected
  ${'1'}        | ${'1'}   | ${true}
  ${'1.0'}      | ${'1'}   | ${true}
  ${'1.0.0'}    | ${'1'}   | ${true}
  ${'1.9.0'}    | ${'1.9'} | ${true}
  ${'1'}        | ${'2'}   | ${false}
  ${'1.9.1'}    | ${'1.9'} | ${false}
  ${'1.9-beta'} | ${'1.9'} | ${false}
`('equals("$a", "$b") === $expected', ({ a, b, expected }) => {
  expect(versioning.equals(a, b)).toBe(expected);
});
```

zharinov replied:
For more examples, take a look at `cargo`...`hex` versioning schemes that already are refactored this way

zharinov replied:
Btw, you don't have to align the columns manually, Prettier will help you

skral replied:
Updated. Please have another look.

skral commented:
- Extracted the pattern and some utility functions into their own files.
- Updated the tests to match the new style.

Please take another look when you get a chance.

skral commented:
Fixed the security warnings.

viceice commented on lib/versioning/rez/transform.ts:
```diff
@@ -0,0 +1,107 @@
+
+  return [versionParts[0], '-' + versionParts[1]];
+}
+export function padZeroes(input: string): string {
```
```suggestion

export function padZeroes(input: string): string {
```

Please add empty newlines between functions

viceice commented on lib/versioning/rez/index.ts:
```diff
@@ -1,59 +1,203 @@
+export const supportsRanges = true;
+export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+
+const equals = (a: string, b: string): boolean => {
```
```suggestion
function equals(a: string, b: string): boolean {
```

Please use function declarations, would make stack traces more readable when unexpected  errors are occuring

skral commented:
@viceice / @zharinov, anything else for me to tackle on this one?

rarkins reviewed (approved):
Versioning on it own won't hurt anybody even if there's an undiscovered bug
